fe/fe-core/src/main/java/com/starrocks/load/routineload/PulsarTaskInfo.java

```diff
@@ -0,0 +1,186 @@
+import com.starrocks.thrift.TFileFormatType;
+import com.starrocks.thrift.TLoadSourceType;
+import com.starrocks.thrift.TPlanFragment;
+import com.starrocks.thrift.TPulsarLoadInfo;
+import com.starrocks.thrift.TRoutineLoadTask;
+import com.starrocks.thrift.TUniqueId;
+
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+
+public class PulsarTaskInfo extends RoutineLoadTaskInfo {
+    private RoutineLoadManager routineLoadManager = GlobalStateMgr.getCurrentState().getRoutineLoadManager();
+
+    private List<String> partitions;
+    private Map<String, Long> initialPositions = Maps.newHashMap();
+
+    public PulsarTaskInfo(UUID id, long jobId, String clusterName, long taskScheduleIntervalMs, long timeToExecuteMs,
+                          List<String> partitions, Map<String, Long> initialPositions) {
+        super(id, jobId, taskScheduleIntervalMs, timeToExecuteMs);
+        this.partitions = partitions;
+        this.initialPositions.putAll(initialPositions);
+    }
+
+    public PulsarTaskInfo(long timeToExecuteMs, PulsarTaskInfo pulsarTaskInfo, Map<String, Long> initialPositions) {
+        super(UUID.randomUUID(), pulsarTaskInfo.getJobId(), pulsarTaskInfo.getTaskScheduleIntervalMs(),
+                timeToExecuteMs, pulsarTaskInfo.getBeId());
+        this.partitions = pulsarTaskInfo.getPartitions();
+        this.initialPositions.putAll(initialPositions);
+    }
+
+    public List<String> getPartitions() {
+        return partitions;
+    }
+
+    public Map<String, Long> getInitialPositions() {
+        return initialPositions;
+    }
+
+    @Override
+    public boolean readyToExecute() throws UserException {
+        RoutineLoadJob routineLoadJob = routineLoadManager.getJob(jobId);
+        if (routineLoadJob == null) {
+            return false;
+        }
+
+        // Got initialPositions, we need to execute even there's no backlogs
+        if (!initialPositions.isEmpty()) {
+            return true;
+        }
+
+        PulsarRoutineLoadJob pulsarRoutineLoadJob = (PulsarRoutineLoadJob) routineLoadJob;
+        Map<String, Long> backlogNums = PulsarUtil.getBacklogNums(pulsarRoutineLoadJob.getServiceUrl(),
+                pulsarRoutineLoadJob.getTopic(), pulsarRoutineLoadJob.getSubscription(),
+                ImmutableMap.copyOf(pulsarRoutineLoadJob.getConvertedCustomProperties()),
+                partitions);
+        for (String partition : partitions) {
+            Long backlogNum = backlogNums.get(partition);
+            if (backlogNum != null && backlogNum > 0) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    // TODO(chen9t) there's no way to find out how many backlogs have been consumed in this round,
+    // the bellowing method will preempt the slots of BEs. So return ture until we find a better way.
+    @Override
+    public boolean isProgressKeepUp(RoutineLoadProgress progress) {
+        // PulsarProgress pProgress = (PulsarProgress) progress;
```

Comment:
How often is the inaccurate backlog returned? We should consider the two sides of the same coin.

Reply:
In my test scenario, it's always not right.